packages/@expo/server/src/index.ts

```diff
@@ -157,148 +141,207 @@ export function createRequestHandler(
-          return contents;
+        try {
+          const html = await getHtml(request, route);
+          return respondHTML(html, route);
+        } catch (error) {
+          // Shows RedBox in development
+          return handleRouteError(error as Error);
         }
-
-        return new Response(contents, { status: 200, headers: { 'Content-Type': 'text/html' } });
       }
     }
 
     // Next, test API routes
-    for (const route of routesManifest.apiRoutes) {
+    for (const route of manifest.apiRoutes) {
       if (!route.namedRegex.test(sanitizedPathname)) {
         continue;
       }
 
-      const func = await getApiRoute(route);
-
-      if (func instanceof Response) {
-        return func;
-      }
-
-      const routeHandler = func?.[request.method];
-      if (!routeHandler) {
-        return new Response('Method not allowed', {
-          status: 405,
-          headers: { 'Content-Type': 'text/plain' },
-        });
-      }
-
-      // Mutate to add the expoUrl object.
-      const params = updateRequestWithConfig(request, route);
-
       try {
-        // TODO: Handle undefined
-        return (await routeHandler(request, params)) as Response;
+        const mod = await getApiRoute(route);
+        return await respondAPI(mod, request, route);
       } catch (error) {
-        if (error instanceof Error) {
-          logApiRouteExecutionError(error);
-        }
-
-        return handleApiRouteError(error as Error);
+        // Shows RedBox in development
+        return handleRouteError(error as Error);
       }
     }
 
     // Finally, test 404 routes
-    for (const route of routesManifest.notFoundRoutes) {
-      if (!route.namedRegex.test(sanitizedPathname)) {
-        continue;
-      }
-
-      // // Mutate to add the expoUrl object.
-      updateRequestWithConfig(request, route);
-
-      // serve a static file
-      const contents = await getHtml(request, route);
+    if (request.method === 'GET' || request.method === 'HEAD') {
+      for (const route of manifest.notFoundRoutes) {
+        if (!route.namedRegex.test(sanitizedPathname)) {
+          continue;
+        }
 
-      // TODO: What's the standard behavior for malformed projects?
-      if (!contents) {
-        return new Response('Not found', {
-          status: 404,
-          headers: { 'Content-Type': 'text/plain' },
-        });
-      } else if (contents instanceof Response) {
-        return contents;
+        try {
+          const contents = await getHtml(request, route);
+          return respondNotFoundHTML(contents, route);
+        } catch {
+          // NOTE(@krystofwoldrich): Should we show a dismissible RedBox in development?
+          // Handle missing/corrupted not found route files
+          continue;
+        }
       }
-
-      return new Response(contents, { status: 404, headers: { 'Content-Type': 'text/html' } });
     }
 
     // 404
-    const response = new Response('Not found', {
+    return new Response('Not found', {
```

Comment:
Future PR/change: We'll have to find a way to mark this for `deployment-worker`. I reckon, we either need another callback function for this, or we have a special `extends Response` markers with a `cause|code` or something?

Alternatively, we could add `status|statusText` to `ExpoError`, throw this, and add boilerplate code to `@expo/server` instead that won't be used in `deployment-worker` to handle this. Not sure if that would work.

Same for the manifest error below, I suppose